kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js

```diff
@@ -283,116 +311,175 @@ define([
-                    });
+        /**
+         *
+         * @param {Object} dataElement
+         */
+        async function doChange(dataElement) {
+            const result = await validate();
+            const changeMsg = {
+                newDisplayValue: undefined,
+                newValue: spec.data.nullValue,
+            };
+
+            if (result.isValid) {
+                const newValue =
+                    result.parsedValue === undefined ? result.value : result.parsedValue;
+                model.value = newValue;
+                if (dataElement) {
+                    model.displayValue = [...model.descriptionFields].reduce((acc, key) => {
+                        acc[key] = dataElement[key];
+                        return acc;
+                    }, {});
+                } else {
+                    model.displayValue = undefined;
                 }
-                channel.emit('validation', {
-                    errorMessage: result.errorMessage,
-                    diagnosis: result.diagnosis,
-                });
+                changeMsg.newDisplayValue = model.displayValue;
+                changeMsg.newValue = newValue;
+            } else if (result.diagnosis === Constants.DIAGNOSIS.REQUIRED_MISSING) {
+                model.value = spec.data.nullValue;
+            }
+            channel.emit('changed', changeMsg);
+            channel.emit('validation', {
+                errorMessage: result.errorMessage,
+                diagnosis: result.diagnosis,
             });
         }
 
         /**
          * Clears the current selection and updates the model.
          */
         function doClear() {
+            $(ui.getElement('input-container.input')).html('');
             model.value = spec.data.nullValue;
             channel.emit('changed', {
+                newDisplayValue: undefined,
                 newValue: spec.data.nullValue,
             });
         }
 
         /**
-         * Formats the display of an object in the dropdown.
-         * id: "data/bulk/wjriehl/subfolder/i_am_a_file.txt"
-         * isFolder: false
-         * mtime: 1508441424000
-         * name: "i_am_a_file.txt"
-         * path: "data/bulk/wjriehl/subfolder/i_am_a_file.txt"
-         * size: 0
-         * text: "data/bulk/wjriehl/subfolder/i_am_a_file.txt"
+         * Formats Staging Area file data for display in the dropdown.
+         * Expects an example fileData object as below:
+         * {
+         *   id: "data/bulk/wjriehl/subfolder/i_am_a_file.txt"
+         *   isFolder: false
+         *   mtime: 1508441424000
+         *   name: "i_am_a_file.txt"
+         *   path: "data/bulk/wjriehl/subfolder/i_am_a_file.txt"
+         *   size: 0
+         *   text: "data/bulk/wjriehl/subfolder/i_am_a_file.txt"
+         * }
+         * @param {Object} fileData
+         * @returns {JQueryElement} a jQuery-wrapped div element
+         */
+        function formatFtpStagingDisplay(fileData) {
+            return $(
```

Comment:
Took this out of the `formatObjectDisplay` function.
Really, the StagingService dropdown could be its own input, or a subclass of this one if we were doing object-oriented things properly.